File: leetcode_learning/code_cpp/134.cpp
//解法一、暴力双循环，依次尝试每个起始位置，看能否走完全程
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        for(int i=0;i<gas.size();i++)
        {
            int cur = i;
            int rest = gas[cur] - cost[cur];
            while(rest>=0)
            {
                cur = cur + 1;
                if(cur==gas.size())
                    cur = 0;
                //这两步可以简化为cur = (cur+1)%gas.size();
                if(cur==i)
                    return i;
                rest += gas[cur] - cost[cur];
            }
        }
        return -1;
    }
};


//解法二、官方解法，需要一点数学证明：如果从x出发无法到达y的下一站，则从x,y之间任意一站(包括x,y)出发，都无法到达下一站
//技术路线：首先检查第0个加油站，并试图判断能否环绕一周；如果不能，就从第一个无法到达的加油站开始继续检查；如果检查到第n个加油站还没有找到符合的，就返回-1。
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int i = 0;
        while (i < gas.size())
        {
            int rest = 0;
            int cnt = 0;//当前起点能够持续走的加油站个数
            while(cnt < gas.size())
            {
                int j = (i + cnt) % gas.size();
                rest += gas[j] - cost[j];
                if (rest<0)
                    break;
                cnt++;
            }
            if (cnt == gas.size())
                return i;
            else
                i = i + cnt + 1;
        }
        return -1;
    }
};

File: leetcode_learning/auto-leetcode/134.加油站/134.py
"""
134.加油站
难度：中等
链接：https://leetcode.cn/problems/gas-station/
"""
from typing import List


class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        # 总油量不够，一定无解
        if sum(gas) < sum(cost):
            return -1

        start = 0  # 起始加油站
        tank = 0    # 当前剩余油量

        for i in range(len(gas)):
            tank += gas[i] - cost[i]
            # 油量不足，i 以及之前的站点都不可能作为起点
            if tank < 0:
                start = i + 1
                tank = 0

        return start


# ====== 错误思路（反例警示）======
# 错误想法：总和>=0时，从净收益最大的站出发即可
# 反例：gas = [5,8,2,8], cost = [6,5,6,6]
#   净收益 = [-1, 3, -4, 2]，总和=0>=0
#   max(净收益)=3 在 index=1，但从 index=1 出发：tank=3 -> 3-4=-1 失败
#   正确答案是 index=3：tank=2 -> 2-1=1 -> 1+3=4 -> 4-4=0 成功
class Solution_Wrong:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        for i in range(len(gas)):
            gas[i] -= cost[i]
        if sum(gas) >= 0:
            return gas.index(max(gas))
        else:
            return -1